README.md
# flutter_rx_stream_builder

A new Flutter package project.

## Getting Started

An rx stream builder widget that is able to pre-populate a flutter `StreamBuilder` with data from an rx stream if the stream is either a value or a replay observable.  For example the RX stream is a `BehaviorSubject` or a `ReplaySubject`.

This will slightly improve the performance as the first frame will be rendered with data rather than waiting for the stream to emit data.

A normal dart stream can also be passed to this widget and it will behave exactly the same way as a normal Flutter `StreamBuilder`

```
Widget build(BuildContext context) {
  return RxStreamBuilder<String>(
      stream: BehaviorSubject<String>.seeded("Hello"),
      builder: (context, snapshot) => Text(snapshot.data),
  );
}
```

For a more complex example see the example project

The below changes the applications theme every two seconds

```
import 'package:flutter/material.dart';
import 'dart:math' as math;
import 'package:rxdart/rxdart.dart';
import 'package:flutter_rx_stream_builder/flutter_rx_stream_builder.dart';

class RandomThemeManager {
  final Subject<ThemeData> _themeData = BehaviorSubject.seeded(
      ThemeData.light().copyWith(primaryColor: Colors.yellow));

  Observable<ThemeData> get themeData => _themeData;

  void initialise() {
    Observable.fromIterable(Iterable.generate(5000, (n) => n))
        .interval(const Duration(seconds: 2))
        .map((_) => ThemeData.light().copyWith(
            primaryColor:
                Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0)
                    .withOpacity(1.0)))
        .listen((theme) => _themeData.add(theme));
  }
}

void main() {
  final themeManager = RandomThemeManager();
  themeManager.initialise();
  runApp(MyApp(themeManager.themeData));
}

class MyApp extends StatelessWidget {
  final Observable<ThemeData> _themeData$;

  const MyApp(this._themeData$, {Key key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return RxStreamBuilder(
      stream: _themeData$,
      builder: (context, snapshot) => MaterialApp(
            title: 'RxStreamBuilder Demo',
            theme: snapshot.data,
            home: MyHomePage(title: 'Rx StreamBuilder Demo Page'),
          ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

CHANGELOG.md
## [2.0.1] - 31/05/2021
* Fixed a null assertion error with ValueStream

## [2.0.0] - 15/04/2021
* Updated to null-safety

## [1.0.3] - 18/12/2019
* Updated to adhere with new names in RxDart

## [1.0.2] - 21/08/2019
* Fixed issue getting initial data of ReplaySubject with no value in value collection

## [1.0.1] - 01/07/2019
* Added ability to define snapshot index in the case where a ReplayObservable is passed in and may have a collection of values accessible

## [1.0.0] - 28/06/2019
* Initial release


lib/src/rx_stream_builder.dart
import 'package:flutter/widgets.dart';
import 'package:rxdart/rxdart.dart';

/// Rx stream builder that will pre-populate the streams initial data if the
/// given stream is an rx observable that holds the streams current value such
/// as a `BehaviorSubject` or a `ReplaySubject`
class RxStreamBuilder<T> extends StatefulWidget {
  /// The asynchronous computation to which this builder is currently connected,
  /// possibly null. When changed, the current summary is updated using
  /// [afterDisconnected], if the previous stream was not null, followed by
  /// [afterConnected], if the new stream is not null.
  final Stream<T> stream;

  /// The build strategy currently used by this builder.
  final AsyncWidgetBuilder<T> builder;

  /// The data that will be used to create the initial snapshot.
  /// Note that the data held with a value / replay  stream (if is is not null)
  /// will take precedence over the data provided.
  final T? initialData;

  /// If the given stream is a ReplayObservable then this zero-based index
  /// will be used to pick the value to use as the initial data.
  /// Defaults to the last emitted value.
  final int? valuesSnapshotIndex;

  /// Creates a new [RxStreamBuilder] that builds itself based on the latest
  /// snapshot of interaction with the specified [stream] and whose build
  /// strategy is given by [builder].
  ///
  /// If the given stream is of a type that stores data like a BehaviorSubject,
  /// ReplaySubject or a variant it's stored data will be used for the inner
  /// StreamBuilder's initial data giving a measurable performance improvement.
  ///
  /// The [initialData] is used to create the initial snapshot.
  ///
  /// The [stream] must not be null.
  /// The [builder] must not be null.
  const RxStreamBuilder({
    required this.stream,
    required this.builder,
    this.initialData,
    this.valuesSnapshotIndex,
    Key? key,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => _RxStreamBuilderState<T>();
}

class _RxStreamBuilderState<T> extends State<RxStreamBuilder<T>> {
  @override
  Widget build(BuildContext context) {
    final initialData = _getInitialData(widget.stream);
    return StreamBuilder(
      initialData: initialData,
      stream: widget.stream,
      builder: widget.builder,
    );
  }

  T? _getInitialData(Stream<T> stream) {
    T? initialData;
    if (widget.initialData != null) {
      initialData = widget.initialData!;
    } else if (stream is ValueStream<T> && stream.hasValue) {
      initialData = stream.value;
    } else if (stream is ReplayStream<T>) {
      if (widget.valuesSnapshotIndex != null &&
          stream.values.length >= widget.valuesSnapshotIndex!) {
        initialData = stream.values.elementAt(widget.valuesSnapshotIndex!);
      } else if (stream.values.isNotEmpty) {
        initialData = stream.values.last;
      }
    }

    return initialData;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;
import 'package:rxdart/rxdart.dart';
import 'package:flutter_rx_stream_builder/flutter_rx_stream_builder.dart';

class RandomThemeManager {
  final Subject<ThemeData> _themeData = BehaviorSubject.seeded(
      ThemeData.light().copyWith(primaryColor: Colors.yellow));

  Stream<ThemeData> get themeData => _themeData;

  void initialise() {
    Stream.fromIterable(Iterable.generate(5000, (n) => n))
        .interval(const Duration(seconds: 2))
        .map((_) => ThemeData.light().copyWith(
            primaryColor:
                Color((math.Random().nextDouble() * 0xFFFFFF).toInt() << 0)
                    .withOpacity(1.0)))
        .listen((theme) => _themeData.add(theme));
  }
}

void main() {
  final themeManager = RandomThemeManager();
  themeManager.initialise();
  runApp(MyApp(themeManager.themeData));
}

class MyApp extends StatelessWidget {
  final Stream<ThemeData> _themeData$;

  const MyApp(this._themeData$) : super();

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return RxStreamBuilder<ThemeData>(
      stream: _themeData$,
      builder: (context, snapshot) => MaterialApp(
        title: 'RxStreamBuilder Demo',
        theme: snapshot.data,
        home: MyHomePage('Rx StreamBuilder Demo Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage(
    this.title,
  ) : super();

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
